Add qr-code generator for external urls for frontends and login

TheMeinerLP opened the pull request:
This pull request includes:

- [ ] breaking changes
- [X] no breaking changes

### Changes made to the repository:

- Add external urls for the qrcode.
- Add qrcode generator.
- Add login option via command
- Add setup for rest api

### Documentation of test results:

Tests works fine.

### Related issues/discussions:

Related/Blocked PR: https://github.com/CloudNetService/CloudNet-v3/pull/696

schucko1337 commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/commands/SetupExternalUrlCommand.java:
```diff
@@ -0,0 +1,69 @@
+
+    this.console.startAnimation(this.animation);
+  }
+
```
Whats this?

schucko1337 commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/commands/SetupExternalUrlCommand.java:
```diff
@@ -0,0 +1,69 @@
+    this.console = console;
+    this.module = module;
+  }
+
```
What happened here?

schucko1337 commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/setup/HttpHostAndPortSetup.java:
```diff
@@ -0,0 +1,85 @@
+  }
+
+  @Override
+  public void handleResults(@NonNull ConsoleSetupAnimation animation) {
```
I don't like the ternary usages in this function. But if @derklaro likes it, it's okay.

schucko1337 commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/v2/V2HttpHandlerQRCode.java:
```diff
@@ -0,0 +1,93 @@
+
+public class V2HttpHandlerQRCode implements HttpHandler {
+
+
```
What happened here?

schucko1337 commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/v2/V2HttpHandlerQRCode.java:
```diff
@@ -0,0 +1,145 @@
+      var sessionValue = RestUtil.first(request.queryParameters().get("session"));
+      if (sessionValue != null) {
+        var token = this.getBearToken(
+          CloudNetDriver.instance().permissionManagement().user(UUID.fromString(sessionValue)));
```
Is it really necessary to retrieve the user by their UUID just to call uniqueId() in line 143?

TheMeinerLP replied:
Yes, of course. The architecture requires it

schucko1337 commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/v2/V2HttpHandlerQRCode.java:
```diff
@@ -0,0 +1,145 @@
+    context.response().status(HttpResponseCode.OK).body(outputStream.toByteArray());
+  }
+
+  public void rebuildQrCodeData() {
```
Please stick to your naming. You used `QrCode` and `QRCode`.

schucko1337 commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/v2/V2HttpHandlerQRCode.java:
```diff
@@ -0,0 +1,145 @@
+    }
+  }
+
+  private void showQRCode(@NotNull HttpContext context, @NotNull String token) throws Exception {
```
Please stick to your naming. You used `QrCode` and `QRCode`.

schucko1337 commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/v2/V2HttpHandlerQRCode.java:
```diff
@@ -0,0 +1,145 @@
+import java.util.concurrent.TimeUnit;
+import org.jetbrains.annotations.NotNull;
+
+public class V2HttpHandlerQRCode extends V2HttpHandler {
```
Please stick to your naming. You used `QrCode` and `QRCode`.

schucko1337 commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/v2/V2HttpHandlerQRCode.java:
```diff
@@ -0,0 +1,145 @@
+      }
+    }
+    var content = JsonDocument.newDocument("externalUrls", externalUrls);
+    try {
```
Please try to generalize this snippet to appear only once in the code.

schucko1337 commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/v2/V2HttpHandlerQRCode.java:
```diff
@@ -0,0 +1,145 @@
+    content.append("name", Node.instance().nodeUniqueId());
+    content.append("token", token);
+    try {
+      BitMatrix matrix = new MultiFormatWriter().encode(content.toString(), BarcodeFormat.QR_CODE, this.width,
```
Please try to generalize this snippet to appear only once in the code.

schucko1337 commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/v2/V2HttpHandlerQRCode.java:
```diff
@@ -0,0 +1,145 @@
+    content.append("name", Node.instance().nodeUniqueId());
+    try {
+      this.byteArrayOutputStream.reset();
+      BitMatrix matrix = new MultiFormatWriter().encode(content.toString(), BarcodeFormat.QR_CODE, this.width,
```
Please try to generalize this snippet to appear only once in the code.

schucko1337 commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/v2/V2HttpHandlerQRCode.java:
```diff
@@ -0,0 +1,145 @@
+    this.authentication.createJwt(user, TimeUnit.MINUTES.toMillis(this.restConfiguration.jwtValidTimeMinutes()));
+  }
+
+  private String getBearToken(PermissionUser user) {
```
Did you mean Bearer Token?

schucko1337 commented on node/src/main/java/eu/cloudnetservice/node/config/HttpHostAndPort.java:
```diff
@@ -0,0 +1,28 @@
+import lombok.NonNull;
+
+/**
+ * Represents an immutable host,port and ssl mapping. A host of this class might be an ipv4/ipv6 address or a domain.
```
```suggestion
 * Represents an immutable host, port, and SSL mapping. A host of this class might be an ipv4/ipv6 address or a domain.
```

schucko1337 commented on node/src/main/java/eu/cloudnetservice/node/config/HttpHostAndPort.java:
```diff
@@ -0,0 +1,28 @@
+ *
+ * @since 4.0
+ */
+public record HttpHostAndPort(@NonNull String host, int port, boolean ssl, boolean ignorePort) {
```
This record represents not only Host and Port. Do you think this is really the correct name for this?

schucko1337 commented on node/src/main/resources/lang/en_US.properties:
```diff
@@ -374,3 +374,13 @@ module-docker-command-remove-collection-property=The {0$property$} {1$value$} wa
+#
+module-rest-qrcode-info=The QR code is generated for the following url:
+module-rest-qrcode-login-info=The QR code is generated for the following url with login token:
+module-rest-qrcode-login-failed=Your given username or password is wrong or not existing
```
```suggestion
module-rest-qrcode-login-failed=Your given username or password is wrong or does not exist
```

schucko1337 commented on node/src/main/resources/lang/en_US.properties:
```diff
@@ -374,3 +374,13 @@ module-docker-command-remove-collection-property=The {0$property$} {1$value$} wa
+module-rest-qrcode-info=The QR code is generated for the following url:
+module-rest-qrcode-login-info=The QR code is generated for the following url with login token:
+module-rest-qrcode-login-failed=Your given username or password is wrong or not existing
+command-http-setup-question-reverseProxy=Is the reset api behind a reverse proxy?
```
```suggestion
command-http-setup-question-reverseProxy=Is the Rest API behind a reverse proxy?
```

schucko1337 commented on node/src/main/resources/lang/en_US.properties:
```diff
@@ -374,3 +374,13 @@ module-docker-command-remove-collection-property=The {0$property$} {1$value$} wa
+module-rest-qrcode-login-info=The QR code is generated for the following url with login token:
+module-rest-qrcode-login-failed=Your given username or password is wrong or not existing
+command-http-setup-question-reverseProxy=Is the reset api behind a reverse proxy?
+command-http-setup-question-hostAndPort=What is the host and port of the reset api?
```
```suggestion
command-http-setup-question-hostAndPort=What is the host and port of the Rest API?
```

schucko1337 commented on node/src/main/resources/lang/en_US.properties:
```diff
@@ -374,3 +374,13 @@ module-docker-command-remove-collection-property=The {0$property$} {1$value$} wa
+module-rest-qrcode-login-failed=Your given username or password is wrong or not existing
+command-http-setup-question-reverseProxy=Is the reset api behind a reverse proxy?
+command-http-setup-question-hostAndPort=What is the host and port of the reset api?
+command-http-setup-question-domain=What is the domain of the reset api?
```
```suggestion
command-http-setup-question-domain=What is the domain of the Rest API?
```

schucko1337 commented on node/src/main/resources/lang/en_US.properties:
```diff
@@ -374,3 +374,13 @@ module-docker-command-remove-collection-property=The {0$property$} {1$value$} wa
+command-http-setup-question-reverseProxy=Is the reset api behind a reverse proxy?
+command-http-setup-question-hostAndPort=What is the host and port of the reset api?
+command-http-setup-question-domain=What is the domain of the reset api?
+command-http-setup-question-ssl=Is the reset api using ssl?
```
```suggestion
command-http-setup-question-ssl=Is the Rest API using SSL?
```

schucko1337 commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/setup/RestSetup.java:
```diff
@@ -0,0 +1,85 @@
+      .key("ssl")
+      .translatedQuestion("command-http-setup-question-ssl")
+      .answerType(QuestionAnswerType.<Boolean>builder().parser(Parsers.bool()).build()).build());
+
```
Remove unnecessary empty lines

schucko1337 commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/commands/RestCommand.java:
```diff
@@ -0,0 +1,154 @@
+    } else {
+      source.sendMessage(I18n.trans("module-rest-qrcode-login-failed"));
+    }
+
```
Remove unnecessary line breaks.

schucko1337 commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/commands/RestCommand.java:
```diff
@@ -0,0 +1,154 @@
+        user = user1;
+      }
+    }
+    if (user != null) {
```
Please use early returns to minimize scope depth.

schucko1337 commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/commands/RestCommand.java:
```diff
@@ -0,0 +1,154 @@
+          stringBuilder.append("http://");
+        }
+        if (externalUrl.ignorePort()) {
+          stringBuilder.append(externalUrl.host());
```
Remove the duplicate `stringBuilder.append(externalUrl.host())`.

schucko1337 commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/commands/RestCommand.java:
```diff
@@ -0,0 +1,154 @@
+        stringBuilder.append("http://");
+      }
+      if (externalUrl.ignorePort()) {
+        stringBuilder.append(externalUrl.host());
```
Remove duplicate `stringBuilder.append(externalUrl.host())`

schucko1337 commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/commands/RestCommand.java:
```diff
@@ -0,0 +1,154 @@
+import java.util.List;
+import lombok.NonNull;
+
+
```
Remove unnecessary line break

schucko1337 commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/commands/RestCommand.java:
```diff
@@ -0,0 +1,154 @@
+
+package eu.cloudnetservice.modules.rest.commands;
+
+
```
Remove unnecessary line break

schucko1337 commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/CloudNetRestModule.java:
```diff
@@ -115,5 +120,11 @@ public void initHttpHandlers() {
+    Node.instance().commandProvider().register(new RestCommand(Node.instance().console(), this));
+  }
+
+  public V2HttpHandlerQRCode getV2HttpHandlerQRCode() {
```
I dont like that this class is used to provide getters. It should be possible to access the V2HttpHandlerQRCode instance when passing it down to RestCommand -> RestSetup.

schucko1337 commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/CloudNetRestModule.java:
```diff
@@ -40,9 +42,12 @@
 
 public final class CloudNetRestModule extends DriverModule {
 
+  private final V2HttpHandlerQRCode v2HttpHandlerQRCode = new V2HttpHandlerQRCode();
```
Move this to initHttpHandlers for a cleaner class structure. (See comment below)

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/CloudNetRestModule.java:
```diff
@@ -40,9 +42,13 @@
+
   @ModuleTask(order = 120, event = ModuleLifeCycle.STARTED)
   public void initHttpHandlers() {
+    V2HttpHandlerQRCode v2HttpHandlerQRCode = new V2HttpHandlerQRCode();
```
Use var

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/commands/RestCommand.java:
```diff
@@ -0,0 +1,121 @@
+    \\/ \\_/\\___||___/\\__|          
+      """;
+  private final String footer = "Rest setup complete!";
+  private final ConsoleSetupAnimation animation = new ConsoleSetupAnimation(this.header, this.footer, "&r> &e");
```
should be static

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/commands/RestCommand.java:
```diff
@@ -0,0 +1,121 @@
+
+  private final V2HttpHandlerQRCode v2HttpHandlerQRCode;
+
+  public RestCommand(Console console, V2HttpHandlerQRCode v2HttpHandlerQRCode) {
```
Nullability annotations

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/commands/RestCommand.java:
```diff
@@ -0,0 +1,121 @@
+  }
+
+  @CommandDescription("Displays the url of the qrcode")
+  @CommandMethod(value = "rest qrcode")
```
```suggestion
  @CommandMethod("rest qrcode")
```

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/commands/RestCommand.java:
```diff
@@ -0,0 +1,121 @@
+        user = user1;
+      }
+    }
+    if (null == user) {
```
This is not how cloud works. Use a parser

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/commands/RestCommand.java:
```diff
@@ -0,0 +1,121 @@
+    this.printExternalUrls(user.uniqueId().toString(), source);
+  }
+
+  private void printExternalUrls(@Nullable String session, @NotNull CommandSource sender) {
```
Dont mix jetbrains and lombok annos

TheMeinerLP replied:
You do it self inside of cloudnet. Please check again own code stylings

0utplay replied:
Can you name a place since it is not intended

TheMeinerLP replied:
<img width="786" alt="image" src="https://user-images.githubusercontent.com/6745190/173209141-5f8482ef-0a10-4bae-a21d-00ea5b1a443f.png">

0utplay replied:
Thats perfectly fine. Papi uses NotNull from jetbrains.

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/commands/RestCommand.java:
```diff
@@ -0,0 +1,121 @@
+        stringBuilder.append("?session=").append(session);
+      }
+      sender.sendMessage(stringBuilder.toString());
+      stringBuilder.setLength(0);
```
Dont see the need of clearing the builder here

TheMeinerLP replied:
See you later

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/util/QRCodeBuilder.java:
```diff
@@ -0,0 +1,69 @@
+import lombok.NonNull;
+import org.jetbrains.annotations.NotNull;
+
+public final class QRCodeBuilder {
```
This builder should be capable of doing more than just applying defaults

TheMeinerLP replied:
Makes more sense to use a builder pattern

0utplay replied:
Dont know what you want to say with that but the builder should allow things like modifying for example: the height, width and external urls

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/util/QRCodeBuilder.java:
```diff
@@ -0,0 +1,69 @@
+        height, Map.of(EncodeHintType.MARGIN, 0));
+      MatrixToImageWriter.writeToStream(matrix, "png", outputStream);
+    } catch (IOException | WriterException e) {
+      LogManager.logger("QRCodeBuilder").severe("Cannot generate qr code", e);
```
Create a private static final logger using `LogManager.logger(Class)`

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/v2/V2HttpHandlerQRCode.java:
```diff
@@ -0,0 +1,73 @@
+    this.authentication.createJwt(user, TimeUnit.MINUTES.toMillis(this.restConfiguration.jwtValidTimeMinutes()));
+  }
+
+  private String getBearerToken(PermissionUser user) {
```
No get/set prefixes

0utplay commented on node/src/main/java/eu/cloudnetservice/node/config/ExternalURLInformation.java:
```diff
@@ -0,0 +1,28 @@
+
+import lombok.NonNull;
+
+/**
```
Describe each parameter in your javadocs

0utplay commented on node/src/main/java/eu/cloudnetservice/node/config/RestConfiguration.java:
```diff
@@ -17,9 +17,11 @@
 
 public record RestConfiguration(
   @NonNull String corsPolicy,
+  @Nullable ExternalURLInformation externalUrl,
```
I dont like that this is the 2. parameter, should be last or second to last

0utplay commented on node/src/main/resources/lang/en_US.properties:
```diff
@@ -374,3 +374,13 @@ module-docker-command-remove-collection-property=The {0$property$} {1$value$} wa
+module-rest-qrcode-info=The QR code is generated for the following url:
+module-rest-qrcode-login-info=The QR code is generated for the following url with login token:
+module-rest-qrcode-login-failed=Your given username or password is wrong or does not exist
+command-http-setup-question-reverseProxy=Is the Reset API behind a reverse proxy?
```
```suggestion
command-http-setup-question-reverseProxy=Is the REST API behind a reverse proxy?
```

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/v2/V2HttpHandlerQRCode.java:
```diff
@@ -0,0 +1,73 @@
+  }
+
+  private String getBearerToken(PermissionUser user) {
+    return this.authentication.generateJwt(user, this.authentication.sessions().get(user.uniqueId().toString()));
```
You could retrieve the user from the session

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/v2/V2HttpHandlerQRCode.java:
```diff
@@ -0,0 +1,73 @@
+  private final ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+
+  public V2HttpHandlerQRCode() {
+    super(null, "GET");
```
Why dont we need a permission here?

TheMeinerLP replied:
This is just a easy way to deliver information. If i set a permission, you need also a auth head. Is not her required

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/commands/RestCommand.java:
```diff
@@ -0,0 +1,121 @@
+    @NonNull @Argument("password") String password) {
+    var users = CloudNetDriver.instance().permissionManagement().usersByName(username);
+    PermissionUser user = null;
+    for (int i = 0; i < users.size() && user == null; i++) {
```
use streams

TheMeinerLP replied:
I prefer faster primitive loops. Same code at original here but just improved.
<img width="766" alt="image" src="https://user-images.githubusercontent.com/6745190/173209107-01d83ed2-aaee-4cc3-9ddb-d8e1caeea905.png">

0utplay replied:
There are some differences here. As said please use a stream here. Its easier to read & understand

0utplay reviewed (changes requested):
There is still work to be done. Your codestyle differs partly strongly from ours. Please have a look at nullability annoations, null-safety and use var where capable.

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/commands/RestCommand.java:
```diff
@@ -0,0 +1,120 @@
+  @CommandDescription("Displays the url of the qrcode")
+  @CommandMethod("rest qrcode")
+  @CommandPermission("cloudnet.command.rest.qrcode")
+  public void showQRCodeUrl(@NonNull CommandSource source, @Flag("appSchema") Boolean appSchema) {
```
`@Nullable` missing at the flag

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/commands/RestCommand.java:
```diff
@@ -0,0 +1,120 @@
+    this.printExternalUrls(user.uniqueId().toString(), source, appSchema);
+  }
+
+  @Parser(name = "username")
+  public @NonNull PermissionUser username(@NonNull CommandContext<?> $, @NonNull Queue<String> input) {
+    var username = input.remove();
+    var users = CloudNetDriver.instance().permissionManagement().usersByName(username);
+    return users.stream().findFirst()
+      .orElseThrow(() -> new ArgumentNotAvailableException(I18n.trans("command-rest-username-not-found")));
+  }
```
```suggestion
@Parser(name = "username")
  public @NonNull PermissionUser username(@NonNull CommandContext<?> $, @NonNull Queue<String> input) {
    var username = input.remove();
    // only retrieve the password without removing it, tricks cloud into thinking the password wasn't parsed yet
    var password = input.element();
    var users = CloudNetDriver.instance().permissionManagement().usersByName(username);
    // filter for any user having the given password
    return users.stream()
      .filter(user -> user.checkPassword(password))
      .findFirst()
      .orElseThrow(() -> new ArgumentNotAvailableException(I18n.trans("command-rest-username-not-found")));
  }
```
Can you please try something like this and remove the password check from the command itself.
Furthermore move the parser method above all other commands

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/commands/RestCommand.java:
```diff
@@ -0,0 +1,120 @@
+  }
+
+  private void printExternalUrls(String session, @NonNull CommandSource sender, Boolean appSchema) {
+    StringBuilder stringBuilder = new StringBuilder();
```
Please use var and add nullability annotations

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/commands/RestCommand.java:
```diff
@@ -0,0 +1,120 @@
+        stringBuilder.append("?session=").append(session);
+      }
+      if (appSchema != null) {
+        stringBuilder.append("&appSchema=").append(true);
```
Why are you asking for a boolean if you set it to true anyways?

TheMeinerLP replied:
If only check if `&appSchema=` added not exactly the value

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/util/QRCodeBuilder.java:
```diff
@@ -0,0 +1,104 @@
+    return this;
+  }
+
+  public QRCodeBuilder withName() {
```
```suggestion
  public QRCodeBuilder withNodeName() {
```

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/util/QRCodeBuilder.java:
```diff
@@ -0,0 +1,104 @@
+  private final JsonDocument content = JsonDocument.newDocument();
+  private boolean appSchema = false;
+
+  public static QRCodeBuilder builder() {
```
Add static builder method to create a copy of another builder

TheMeinerLP replied:
Makes which sense ?

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/util/QRCodeBuilder.java:
```diff
@@ -0,0 +1,104 @@
+        height, Map.of(EncodeHintType.MARGIN, 0));
+      MatrixToImageWriter.writeToStream(matrix, "png", outputStream);
+    } catch (IOException | WriterException e) {
+      LogManager.logger(QRCodeBuilder.class).severe("Cannot generate qr code", e);
```
Create a private static final logger

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/util/QRCodeBuilder.java:
```diff
@@ -0,0 +1,104 @@
+    return this;
+  }
+
+  public QRCodeBuilder withAppSchema() {
```
Might want to add a method consuming a boolean to set the schema

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/util/QRCodeBuilder.java:
```diff
@@ -0,0 +1,104 @@
+    }
+  }
+
+  private String buildAppSchema() {
```
Nullability...

0utplay commented on node/src/main/java/eu/cloudnetservice/node/config/ExternalURLInformation.java:
```diff
@@ -0,0 +1,41 @@
+
+  public String toUrl() {
+    StringBuilder sb = new StringBuilder();
+    if (ssl()) {
```
use this accessor

0utplay reviewed (changes requested):
In some places nullability annotations are missing as well as the use of var. Please check the whole PR for these

TheMeinerLP commented:
Feature complete from side

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/commands/RestCommand.java:
```diff
@@ -0,0 +1,129 @@
+  }
+
+  @Parser(name = "username")
+  public @NonNull PermissionUser username(@NonNull CommandContext<?> $, @NonNull Queue<String> input) {
```
Move the parser up to the ctor

TheMeinerLP replied:
What you meaning with ctor ?

0utplay replied:
var?

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/commands/RestCommand.java:
```diff
@@ -0,0 +1,129 @@
+        stringBuilder.append("?session=").append(session);
+      }
+      if (appSchema != null) {
+        stringBuilder.append("&appSchema=").append(true);
```
That does not make sense

TheMeinerLP replied:
Need be fixed, is a mistake of me. I add the used boolean

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/util/ExternalURLBuilder.java:
```diff
@@ -0,0 +1,42 @@
+
+public final class ExternalURLBuilder {
+
+  public static List<ExternalURLInformation> buildExternalURLs() {
```
@NonNull

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/util/QRCodeBuilder.java:
```diff
@@ -0,0 +1,119 @@
+
+  private static final Logger LOGGER = LogManager.logger(QRCodeBuilder.class);
+
+  public static QRCodeBuilder builder() {
```
@NonNull

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/util/QRCodeBuilder.java:
```diff
@@ -0,0 +1,119 @@
+
+  public void build(@NonNull OutputStream outputStream) {
+    try {
+      BitMatrix matrix = new MultiFormatWriter().encode(
```
var

0utplay commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/util/QRCodeBuilder.java:
```diff
@@ -0,0 +1,119 @@
+    sb.append("externalUrls=");
+    sb.append(
+      URLEncoder.encode(
+        Joiner.on(',').join(ExternalURLBuilder.buildExternalURLs().stream().map(ExternalURLInformation::toUrl).collect(
```
.stream()
.map()
.collect

0utplay reviewed (changes requested):
Recheck usages of var

schucko1337 commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/util/QRCodeBuilder.java:
```diff
@@ -0,0 +1,117 @@
+          .collect(Collectors.joining(",")), StandardCharsets.UTF_8));
+    return sb.toString();
+  }
+
```
Remove blank lines

schucko1337 commented on modules/rest/src/main/java/eu/cloudnetservice/modules/rest/util/ExternalURLBuilder.java:
```diff
@@ -0,0 +1,43 @@
+    }
+    return externalUrls;
+  }
+
```
Remove blank line

derklaro committed:
[ci skip] build: prepare for next development iteration

derklaro committed:
chore: use sponge download api instead of hardcoded urls (#771)

### Motivation
Sponge provides a downloads api which allows to query the latest version of a sponge artifact. This is much nicer and removes the need to manually ensure that the download url of the sponge versions are always up-to-date.

### Modification
Remove hardcoded sponge download urls from versions.json and replaced them with an install step which fetches the latest version from the sponge downloads api.

### Result
Easy ensure that the downloaded version is always the latest one instead of relying on manual updates.
